design.md

```diff
@@ -0,0 +1,84 @@
+# Bazel Module (bzlmod) Support in Renovate
+
+```yaml
+author: Chuck Grindel<chuck.grindel@gmail.com>
+status: To be reviewed
+created: 2023-04-10
+```
+
+## Abstract
+
+[Renovate] is a popular service for updating external dependencies. It [currently supports] updating
+Bazel dependencies described in [WORKSPACE-managed repositories]. With the introduction of [Bazel
+modules], [Renovate] needs to be modified to support the new external dependency mechanism.
+
+## Requirements
+
+- External Dependency Discovery
+  - Support extracting external dependency information from `MODULE.bazel` files.
+    - Read external dependency information from [bazel_dep] declarations.
+    - Support module overrides.
+      - [Single version override]
+      - [Multiple version override]
+      - [Non-registry overrides]
+- Registry Discovery
+  - Support parsing of `.bazelrc` and files imported in the `.bazelrc` looking for [--registry]
+    flags.
+  - If no registry values are detected, default to the [Bazel Central Registry].
+- Bazel Module Release Detection
+  - Detect new module versions when they are introduced to a module registry (e.g., [Bazel Central
+    Registry]).
+    - Triggering updates based upon GitHub releases is problematic as the update to a registry is a
+      separate step that may not complete for minutes to hours (i.e., human approval of a pull
+      request).
+- Versioning
+  - Support [Bazel module version formats] and [compatibility level].
+  - Support [yanked versions]. A version can be yanked by a maintainer when it should be avoided.
+  - Support different version resolution logic for libraries (i.e., is depended upon by other
```

Comment:
Was a decision made here or is this still up for discussion? IMO, if I'm understanding correctly, this seems like a much bigger project that shouldn't have a bazel specific solution?

Reply:
There was [a Slack discussion](https://bazelbuild.slack.com/archives/C014RARENH0/p1674838476782969). There are tradeoffs either way. 
- If a library keeps its dependencies at the maximum viable version, it forces clients of the library to use a version later than their declared version.
- If a library does not use the maximum viable version, there is a possibility that security and bug fixes will not be picked up.

The only rub I see with library repositories is that we would never upgrade runtime dependencies. Renovate would only propose upgrades for development dependencies. Library maintainers will manually upgrade their runtime dependencies when they need features from a later version of the dependency.

@brentleyjones I believe you were interested in seeing library repositories use the minimum viable version for runtime dependencies. Do you have any thoughts on how this should be codified in the Renovate manager?

Reply:
The slack discussion is what I was basing this on.

I think if a library wants such restrictions on renovate then it should configure renovate to do that. That's already supported by renovate and I don't think needs a bazel specific solution. Or is there a feature in bzlmod that I'm unaware of (which is very possible 😅)?

That's what things like [matchupdatetypes](https://docs.renovatebot.com/configuration-options/#matchupdatetypes), [vulnerabilityAlerts](https://docs.renovatebot.com/configuration-options/#vulnerabilityalerts), [dependency-pinning](https://docs.renovatebot.com/dependency-pinning/) etc are for. I don't think we need another one.